File: python/course-repo/server/webhook_receiver.py
"""Stdlib HTTP webhook receiver used as the webhook receiver in this tutorial.

Endpoints:
  POST /hooks      - Receives a webhook delivery. Records it. Returns 200.
                     Honors an optional Idempotency-Key header (used by Module 02).
                     Randomly returns 500 based on the failure rate set in the
                     Control Panel (read from /tmp/failure_rate).
  GET  /_received  - Returns request and processed-delivery counts for inspection.
  POST /_reset     - Clears recorded state. Used between checks.

No external dependencies - intentionally stdlib-only so it works in any sandbox.
"""

import json
import random
import threading
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, HTTPServer

FAILURE_RATE_FILE = "/tmp/failure_rate"


def _read_failure_rate() -> int:
    """Read the current failure rate (0-100) from the shared file."""
    try:
        with open(FAILURE_RATE_FILE, "r") as f:
            return int(f.read().strip())
    except (FileNotFoundError, ValueError):
        return 0

_received: list[dict] = []
_request_count = 0
_deduped_count = 0
_lock = threading.Lock()


class EchoHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        return  # silence default access logs

    def do_POST(self) -> None:
        if self.path == "/hooks":
            self._handle_hook()
        elif self.path == "/_reset":
            self._handle_reset()
        else:
            self._respond(404, {"error": "not found"})

    def do_GET(self) -> None:
        if self.path == "/":
            self._handle_dashboard()
        elif self.path == "/_received":
            self._handle_received()
        else:
            self._respond(404, {"error": "not found"})

    def _handle_dashboard(self) -> None:
        """Auto-refreshing HTML view of /_received. Polls every 2s via fetch."""
        html = """<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<title>Webhook receiver</title>
<style>
  body { background: #1a1a2e; color: #e2e8f0; font-family: ui-monospace, "SF Mono", monospace; padding: 1em; margin: 0; }
  header { color: #b794f6; font-weight: 600; margin-bottom: 0.6em; }
  pre { white-space: pre-wrap; margin: 0; font-size: 13px; line-height: 1.45; }
  .meta { color: #a0aec0; font-size: 11px; margin-top: 0.4em; }
</style>
</head>
<body>
<header>Webhook receiver &mdash; live</header>
<pre id="data">loading...</pre>
<div class="meta">auto-refreshes every 2s</div>
<script>
async function refresh() {
  try {
    const r = await fetch('/_received');
    const j = await r.json();
    document.getElementById('data').textContent = JSON.stringify(j, null, 2);
  } catch (e) {
    document.getElementById('data').textContent = 'error: ' + e;
  }
}
refresh();
setInterval(refresh, 2000);
</script>
</body>
</html>
"""
        body = html.encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _handle_hook(self) -> None:
        global _request_count, _deduped_count

        length = int(self.headers.get("Content-Length", "0"))
        raw = self.rfile.read(length) if length > 0 else b"{}"
        try:
            body = json.loads(raw)
        except json.JSONDecodeError:
            self._respond(400, {"error": "invalid JSON"})
            return

        # Check failure rate from Control Panel — randomly return 500.
        failure_rate = _read_failure_rate()
        if failure_rate > 0 and random.randint(1, 100) <= failure_rate:
            self._respond(500, {"error": "simulated failure", "failure_rate": failure_rate})
            return

        idem_key = self.headers.get("Idempotency-Key")

        with _lock:
            _request_count += 1

            # Idempotency dedup (exercised in Module 02; harmless in Module 01).
            if idem_key:
                for prior in _received:
                    if prior.get("idempotency_key") == idem_key:
                        _deduped_count += 1
                        self._respond(200, {"status": "ok", "deduped": True})
                        return

            _received.append({
                "received_at": datetime.now(timezone.utc).isoformat(),
                "body": body,
                "idempotency_key": idem_key,
            })

        self._respond(200, {"status": "ok", "deduped": False})

    def _handle_received(self) -> None:
        with _lock:
            self._respond(200, {
                "received_count": _request_count,
                "processed_count": len(_received),
                "deduped_count": _deduped_count,
                # Backward-compatible alias used by the lessons.
                "count": len(_received),
                "deliveries": list(_received),
            })

    def _handle_reset(self) -> None:
        global _request_count, _deduped_count

        with _lock:
            _received.clear()
            _request_count = 0
            _deduped_count = 0
        self._respond(200, {"status": "reset"})

    def _respond(self, code: int, payload: dict) -> None:
        # indent=2 so the echo tab is readable without toggling Pretty-print.
        body = json.dumps(payload, indent=2).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)


def main() -> None:
    server = HTTPServer(("0.0.0.0", 9000), EchoHandler)
    print("Webhook receiver listening on :9000")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: typescript/course-repo/server/webhookReceiver.ts
import http from 'http';
import { URL } from 'url';

interface Delivery {
  receivedAt: string;
  idempotencyKey: string | null;
  body: unknown;
}

const state = {
  receivedCount: 0,
  processedCount: 0,
  dedupedCount: 0,
  throttledCount: 0,
  rateLimit: 0,
  windowTimestamps: [] as number[],
  seenKeys: new Set<string>(),
  deliveries: [] as Delivery[],
};

function resetState() {
  state.receivedCount = 0;
  state.processedCount = 0;
  state.dedupedCount = 0;
  state.throttledCount = 0;
  state.rateLimit = 0;
  state.windowTimestamps = [];
  state.seenKeys.clear();
  state.deliveries = [];
}

function isRateLimited(): boolean {
  if (state.rateLimit === 0) return false;
  const now = Date.now();
  const cutoff = now - 1000;
  state.windowTimestamps = state.windowTimestamps.filter(t => t > cutoff);
  if (state.windowTimestamps.length >= state.rateLimit) {
    return true;
  }
  state.windowTimestamps.push(now);
  return false;
}

function readBody(req: http.IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    req.on('data', c => chunks.push(c));
    req.on('end', () => resolve(Buffer.concat(chunks).toString()));
    req.on('error', reject);
  });
}

function statsJson(): object {
  return {
    received_count: state.receivedCount,
    processed_count: state.processedCount,
    deduped_count: state.dedupedCount,
    throttled_count: state.throttledCount,
    rate_limit: state.rateLimit,
    count: state.processedCount,
    deliveries: state.deliveries.map(d => ({
      received_at: d.receivedAt,
      idempotency_key: d.idempotencyKey,
      body: d.body,
    })),
  };
}

const HTML_PAGE = `<!DOCTYPE html>
<html><head><meta charset="utf-8">
<meta http-equiv="refresh" content="2">
<title>Webhook Receiver</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #e2e8f0; padding: 2rem; }
h1 { color: #7aa2ff; }
.stat { display: inline-block; margin: 0.5rem 1rem 0.5rem 0; padding: 0.5rem 1rem;
        background: #252540; border-radius: 4px; }
.stat .label { color: #a0aec0; font-size: 0.85em; }
.stat .value { font-size: 1.8em; font-weight: bold; color: #9ae6b4; }
.stat .value.red { color: #fc8181; }
pre { background: #252540; padding: 1rem; border-radius: 4px; overflow-x: auto; }
</style>
</head>
<body>
<h1>Webhook Receiver</h1>
<p style="color:#a0aec0">Auto-refreshes every 2 seconds.</p>
<div id="stats"></div>
<pre id="json"></pre>
<script>
fetch('/_received').then(r=>r.json()).then(d=>{
  document.getElementById('json').textContent = JSON.stringify(d, null, 2);
  const stats = document.getElementById('stats');
  const items = [
    ['Received', d.received_count, false],
    ['Processed', d.processed_count, false],
    ['Deduped', d.deduped_count, false],
    ['Throttled (429)', d.throttled_count, d.throttled_count > 0],
    ['Rate limit (req/s)', d.rate_limit || 'off', false],
  ];
  stats.innerHTML = items.map(([l,v,r]) =>
    '<div class="stat"><div class="label">'+l+'</div><div class="value'+(r?' red':'')+'">'
    +v+'</div></div>').join('');
});
</script>
</body></html>`;

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url ?? '/', `http://${req.headers.host}`);

  if (req.method === 'POST' && url.pathname === '/hooks') {
    state.receivedCount++;

    if (isRateLimited()) {
      state.throttledCount++;
      res.writeHead(429, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'Too Many Requests' }));
      return;
    }

    let body: unknown = {};
    try {
      const raw = await readBody(req);
      body = JSON.parse(raw);
    } catch { /* malformed body; accept anyway */ }

    const idempotencyKey = req.headers['idempotency-key'] as string | undefined ?? null;

    if (idempotencyKey !== null && state.seenKeys.has(idempotencyKey)) {
      state.dedupedCount++;
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ deduped: true, idempotencyKey }));
      return;
    }

    if (idempotencyKey !== null) {
      state.seenKeys.add(idempotencyKey);
    }

    state.processedCount++;
    state.deliveries.push({
      receivedAt: new Date().toISOString(),
      idempotencyKey,
      body,
    });

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true, processed: state.processedCount }));
    return;
  }

  if (req.method === 'POST' && url.pathname === '/_reset') {
    resetState();
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ reset: true }));
    return;
  }

  if (req.method === 'POST' && url.pathname === '/_rate_limit') {
    const limit = parseInt(url.searchParams.get('limit') ?? '0', 10);
    state.rateLimit = isNaN(limit) ? 0 : limit;
    state.windowTimestamps = [];
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ rateLimit: state.rateLimit }));
    return;
  }

  if (url.pathname === '/_received') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(statsJson(), null, 2));
    return;
  }

  if (req.method === 'GET' && (url.pathname === '/' || url.pathname === '')) {
    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(HTML_PAGE);
    return;
  }

  res.writeHead(404);
  res.end('Not found');
});

const PORT = 9000;
server.listen(PORT, '0.0.0.0', () => {
  console.log(`Webhook receiver listening on :${PORT}`);
});
